# Data Processing (First 3 Genres)
Ashley Feiler, aef56@pitt.edu

### Table of Contents
- [Children's Literature](#Children's-Literature-Data) - Processes the data for the Children's Literature Reviews
    - [Loading Data](#Children's-Literature-Reviews) - Loads review, book, genre, and author info and merges into one dataframe 
    - [Language Filtering](#Language-Filtering) - Subsets data to just English language data
    - [Formatting](#Formatting) - Cuts down dataframe columns and removes rows with empty review data
- [Poetry](#Poetry-Data)  - Processes the data for the Poetry Reviews following the same format as above
- [Comics/Graphic Novels](#Comics/Graphic-Novels-Data) - Processes the data for the Comics/Graphic Novels Reviews following the same format as above
- [Data Processing Continued](#Data-Processing-Continued) - Provides additional info on processing remaining genres

## Imports

In [1]:
#Turn off pretty printing
%pprint

#Import packages
import pandas as pd
import os
import pickle

Pretty printing has been turned OFF


# Children's Literature Data

## Children's Literature Reviews
I began by reading in the JSON file with review data and converting to DataFrame to prep for analysis. By focusing on only one genre (Children's Literature), I can explore the content and structure of the data at a more manageable scale, and then apply the successful methods to the other genre data sets.

In [2]:
df = pd.read_json('/Users/ashleyfeiler/Documents/data_science/Goodreads-Genre-Reviews-Analysis/data/Reviews/excerpts_shuf/goodreads_reviews_children_shuf.json', lines=True)

In [3]:
df.head()

user_id   book_id  \
0  a2d6dd1685e5aa0a72c9410f8f55e056  17165891   
1  686b174e58665fe35cd89c64c0607992  29064279   
2  e5b6484501d2667dddf7df4d588311de      3636   
3  254f6270f43e504826a2cbac6ea93791  17835181   
4  fd831e16ac0af21a5557a6e1ce156edf     93531   

                          review_id  rating  \
0  dd6c3698b614b747d6fbd01e65186e34       0   
1  c1c06eeb894e4307875165f4fe8d2f98       4   
2  31a18e28070f4b2cb2139ea92c16c9c5       5   
3  0dfb68d65a67d55e3377641d6cdc1027       3   
4  00a7c13cbe5775e7fb5a40fba5d5a64b       4   

                                         review_text  \
0                                                  O   
1  I liked the illustrations, which are are - wel...   
2                    The best book I've read so far.   
3       3.5 stars \n I enjoyed this book.. Fun read.   
4  Pearl S. Buck tells a touching and memorable s...   

                       date_added                    date_updated  \
0  Fri Oct 11 07:22:22 -0700 2013  Fri Oct 11 07:22:49 -0700 2013   
1  Wed Jan 25 18:06:11 -0800 2017  Wed Jan 25 18:15:49 -0800 2017   
2  Tue Oct 06 16:06:04 -0700 2009  Fri Oct 09 14:07:44 -0700 2009   
3  Thu Dec 26 15:05:48 -0800 2013  Tue Jan 14 16:01:05 -0800 2014   
4  Wed Jan 09 11:05:20 -0800 2013  Wed Jan 09 11:57:47 -0800 2013   

                          read_at                      started_at  n_votes  \
0                                                                        0   
1  Sun Jan 01 00:00:00 -0800 2017                                        0   
2  Fri Oct 09 00:00:00 -0700 2009                                        0   
3  Sat Dec 28 00:00:00 -0800 2013  Fri Dec 27 00:00:00 -0800 2013        0   
4  Sat Dec 01 00:00:00 -0800 2012                                        0   

   n_comments  
0           0  
1           0  
2           0  
3           0  
4           0

In [4]:
#Saves a small portion of the original Children's Literature data to share as a sample
#children_share = df.head()

#f = open('data/genre_share/children_share.pkl', 'wb')
#pickle.dump(children_share, f)
#f.close() 

In [5]:
df.shape

(5000, 11)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   user_id       5000 non-null   object
 1   book_id       5000 non-null   int64 
 2   review_id     5000 non-null   object
 3   rating        5000 non-null   int64 
 4   review_text   5000 non-null   object
 5   date_added    5000 non-null   object
 6   date_updated  5000 non-null   object
 7   read_at       5000 non-null   object
 8   started_at    5000 non-null   object
 9   n_votes       5000 non-null   int64 
 10  n_comments    5000 non-null   int64 
dtypes: int64(4), object(7)
memory usage: 429.8+ KB


Due to the overwhelmingly large size of these files, I took samples of 5000 entries from each genre file (8) through the command line in order to avoid loading a file larger than my computer could handle. This shows that there all 5000 rows are accounted for! I will keep an eye on this number as I add to the DataFrame to make sure I'm not accidentally adding/removing anything. There are no NaN values, which is great, but there are lot of columns whose data is unecessary for this product, so let's trim that down to the essentials. 

In [7]:
reviews_df = df[['user_id', 'book_id', 'review_id', 'rating', 'review_text']]

In [8]:
reviews_df.head()

user_id   book_id  \
0  a2d6dd1685e5aa0a72c9410f8f55e056  17165891   
1  686b174e58665fe35cd89c64c0607992  29064279   
2  e5b6484501d2667dddf7df4d588311de      3636   
3  254f6270f43e504826a2cbac6ea93791  17835181   
4  fd831e16ac0af21a5557a6e1ce156edf     93531   

                          review_id  rating  \
0  dd6c3698b614b747d6fbd01e65186e34       0   
1  c1c06eeb894e4307875165f4fe8d2f98       4   
2  31a18e28070f4b2cb2139ea92c16c9c5       5   
3  0dfb68d65a67d55e3377641d6cdc1027       3   
4  00a7c13cbe5775e7fb5a40fba5d5a64b       4   

                                         review_text  
0                                                  O  
1  I liked the illustrations, which are are - wel...  
2                    The best book I've read so far.  
3       3.5 stars \n I enjoyed this book.. Fun read.  
4  Pearl S. Buck tells a touching and memorable s...

Much better! Now this data contains the book IDs, but I'd like to be able to identify the books by title/author and have some basic information such as page count. This book-specific data is in another JSON file, so let's load that into another DataFrame.

## Children's Literature Book Info

In [9]:
books = pd.read_json('/Users/ashleyfeiler/Documents/data_science/Goodreads-Genre-Reviews-Analysis/data/Books/goodreads_books_children.json', lines=True)

In [10]:
books.head()

isbn  text_reviews_count    series country_code language_code  \
0  1599150603                   7        []           US                 
1  1934876569                   6  [151854]           US                 
2  0590417010                 193        []           US           eng   
3  0915190575                   4        []           US                 
4  1416904999                   4        []           US                 

                                     popular_shelves asin is_ebook  \
0  [{'count': '56', 'name': 'to-read'}, {'count':...         false   
1  [{'count': '515', 'name': 'to-read'}, {'count'...         false   
2  [{'count': '450', 'name': 'to-read'}, {'count'...         false   
3  [{'count': '8', 'name': 'to-read'}, {'count': ...         false   
4  [{'count': '8', 'name': 'to-read'}, {'count': ...         false   

   average_rating kindle_asin  ... publication_month edition_information  \
0            4.13  B00DU10PUG  ...                 9                       
1            4.22              ...                 3                       
2            4.43  B017RORXNI  ...                 9                       
3            4.29              ...                                         
4            3.57              ...                 6                       

  publication_year                                                url  \
0             2006  https://www.goodreads.com/book/show/287141.The...   
1             2009  https://www.goodreads.com/book/show/6066812-al...   
2             1995  https://www.goodreads.com/book/show/89378.Dog_...   
3                   https://www.goodreads.com/book/show/3209312-mo...   
4             2005  https://www.goodreads.com/book/show/1698376.Wh...   

                                           image_url  book_id ratings_count  \
0  https://s.gr-assets.com/assets/nophoto/book/11...   287141            46   
1  https://images.gr-assets.com/books/1316637798m...  6066812            98   
2  https://images.gr-assets.com/books/1360057676m...    89378          1331   
3  https://s.gr-assets.com/assets/nophoto/book/11...  3209312            11   
4  https://s.gr-assets.com/assets/nophoto/book/11...  1698376            23   

   work_id                                              title  \
0   278578                      The Aeneid for Boys and Girls   
1   701117  All's Fairy in Love and War (Avalon: Web of Ma...   
2    86259                                         Dog Heaven   
3  3242879  Moths and Mothers, Feathers and Fathers: A Sto...   
4  1695373                                    What Do You Do?   

                                title_without_series  
0                      The Aeneid for Boys and Girls  
1  All's Fairy in Love and War (Avalon: Web of Ma...  
2                                         Dog Heaven  
3  Moths and Mothers, Feathers and Fathers: A Sto...  
4                                    What Do You Do?  

[5 rows x 29 columns]

In [11]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124082 entries, 0 to 124081
Data columns (total 29 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   isbn                  124082 non-null  object 
 1   text_reviews_count    124082 non-null  int64  
 2   series                124082 non-null  object 
 3   country_code          124082 non-null  object 
 4   language_code         124082 non-null  object 
 5   popular_shelves       124082 non-null  object 
 6   asin                  124082 non-null  object 
 7   is_ebook              124082 non-null  object 
 8   average_rating        124082 non-null  float64
 9   kindle_asin           124082 non-null  object 
 10  similar_books         124082 non-null  object 
 11  description           124082 non-null  object 
 12  format                124082 non-null  object 
 13  link                  124082 non-null  object 
 14  authors               124082 non-null  object 
 15  

This DataFrame once again has no NaN values (nice!), and while it does have a much higher number of samples (124082) than the reviews DataFrame, this makes sense because I loaded the entire file instead of a sample. This is because in order to merget the data from each DataFrame, I will need matching book IDs, and it's impossible to know which book IDs I might need for my sample of reviews. 

This is yet again a ton of info, so let's trim it down again to data I might need.

In [12]:
books_df = books[['text_reviews_count', 'series', 'country_code', 'language_code', 'is_ebook', 'average_rating', 'description', 'format', 'authors', 'publisher', 'num_pages', 'edition_information', 'publication_year', 'book_id', 'ratings_count', 'title', 'title_without_series']]

In [13]:
books_df.head()

text_reviews_count    series country_code language_code is_ebook  \
0                   7        []           US                  false   
1                   6  [151854]           US                  false   
2                 193        []           US           eng    false   
3                   4        []           US                  false   
4                   4        []           US                  false   

   average_rating                                        description  \
0            4.13  Relates in vigorous prose the tale of Aeneas, ...   
1            4.22  To Kara's astonishment, she discovers that a p...   
2            4.43  In Newbery Medalist Cynthia Rylant's classic b...   
3            4.29                                                      
4            3.57  WHAT DO YOU DO?\nA hen lays eggs...\nA cow giv...   

       format                                            authors  \
0   Paperback             [{'author_id': '3041852', 'role': ''}]   
1   Paperback               [{'author_id': '19158', 'role': ''}]   
2   Hardcover                [{'author_id': '5411', 'role': ''}]   
3              [{'author_id': '589328', 'role': ''}, {'author...   
4  Board Book              [{'author_id': '169159', 'role': ''}]   

              publisher num_pages edition_information publication_year  \
0  Yesterday's Classics       162                                 2006   
1            Seven Seas       216                                 2009   
2        Blue Sky Press        40                                 1995   
3                                                                        
4          Little Simon        24                                 2005   

   book_id  ratings_count                                              title  \
0   287141             46                      The Aeneid for Boys and Girls   
1  6066812             98  All's Fairy in Love and War (Avalon: Web of Ma...   
2    89378           1331                                         Dog Heaven   
3  3209312             11  Moths and Mothers, Feathers and Fathers: A Sto...   
4  1698376             23                                    What Do You Do?   

                                title_without_series  
0                      The Aeneid for Boys and Girls  
1  All's Fairy in Love and War (Avalon: Web of Ma...  
2                                         Dog Heaven  
3  Moths and Mothers, Feathers and Fathers: A Sto...  
4                                    What Do You Do?

Let's explore some of these columns.

In [14]:
books_df.country_code.value_counts()

US    124082
Name: country_code, dtype: int64

Looks like every book has the same country code - I can probably leave that column out.

In [15]:
books_df.language_code.value_counts()

         72496
eng      30253
en-US     6197
en-GB     2071
spa       1501
         ...  
egy          1
aze          1
mon          1
vls          1
nub          1
Name: language_code, Length: 90, dtype: int64

This will be important to consider - to scale down the data, I will likely restrict the data to only reviews of books written in English (more on this later!)

In [16]:
books_df.format.value_counts()

Hardcover                   51159
Paperback                   38880
                            21039
Board Book                   3214
ebook                        2423
                            ...  
pop up book                     1
miekka (paperback)              1
Klappenbroschur                 1
three books in boxed set        1
Board Book Periodical           1
Name: format, Length: 209, dtype: int64

209 different formats of books? I have no idea what those could be, and the book format likely doesn't influence the reviews that significantly, so I can probably leave out this column as well.

In [17]:
books_df.is_ebook.value_counts()

false    114606
true       9476
Name: is_ebook, dtype: int64

There are significantly more physical books represented in this data than ebooks.

In [18]:
books_df.loc[books_df['book_id'] == 23310161]

text_reviews_count    series country_code language_code is_ebook  \
17550                1231  [751058]           US           eng    false   

       average_rating                                        description  \
17550            4.43  The companion to the #1 blockbuster bestseller...   

          format                                            authors  \
17550  Hardcover  [{'author_id': '6561846', 'role': ''}, {'autho...   

            publisher num_pages edition_information publication_year  \
17550  Philomel Books        36                                 2015   

        book_id  ratings_count                          title  \
17550  23310161           8924  The Day the Crayons Came Home   

                title_without_series  
17550  The Day the Crayons Came Home

Indexing seems to work!
Now that I've explored the data a bit more, I'm going to remove a few more columns.

In [19]:
books_df = books[['text_reviews_count', 'series', 'language_code', 'is_ebook', 'average_rating', 'description', 'authors', 'publisher', 'num_pages', 'edition_information', 'publication_year', 'book_id', 'ratings_count', 'title', 'title_without_series']]

In [20]:
books_df.head()

text_reviews_count    series language_code is_ebook  average_rating  \
0                   7        []                  false            4.13   
1                   6  [151854]                  false            4.22   
2                 193        []           eng    false            4.43   
3                   4        []                  false            4.29   
4                   4        []                  false            3.57   

                                         description  \
0  Relates in vigorous prose the tale of Aeneas, ...   
1  To Kara's astonishment, she discovers that a p...   
2  In Newbery Medalist Cynthia Rylant's classic b...   
3                                                      
4  WHAT DO YOU DO?\nA hen lays eggs...\nA cow giv...   

                                             authors             publisher  \
0             [{'author_id': '3041852', 'role': ''}]  Yesterday's Classics   
1               [{'author_id': '19158', 'role': ''}]            Seven Seas   
2                [{'author_id': '5411', 'role': ''}]        Blue Sky Press   
3  [{'author_id': '589328', 'role': ''}, {'author...                         
4              [{'author_id': '169159', 'role': ''}]          Little Simon   

  num_pages edition_information publication_year  book_id  ratings_count  \
0       162                                 2006   287141             46   
1       216                                 2009  6066812             98   
2        40                                 1995    89378           1331   
3                                                 3209312             11   
4        24                                 2005  1698376             23   

                                               title  \
0                      The Aeneid for Boys and Girls   
1  All's Fairy in Love and War (Avalon: Web of Ma...   
2                                         Dog Heaven   
3  Moths and Mothers, Feathers and Fathers: A Sto...   
4                                    What Do You Do?   

                                title_without_series  
0                      The Aeneid for Boys and Girls  
1  All's Fairy in Love and War (Avalon: Web of Ma...  
2                                         Dog Heaven  
3  Moths and Mothers, Feathers and Fathers: A Sto...  
4                                    What Do You Do?

Now that I have the preliminary information I want, I'll use pandas' `.merge()` function to combine the review and book data, which will automatically align based off of their shared book_id values.

In [21]:
children_df = pd.merge(reviews_df, books_df)

In [22]:
children_df.head()

user_id   book_id  \
0  a2d6dd1685e5aa0a72c9410f8f55e056  17165891   
1  6adb9b5f52e7a89b820d8afcb6329b0e  17165891   
2  686b174e58665fe35cd89c64c0607992  29064279   
3  e5b6484501d2667dddf7df4d588311de      3636   
4  f59779e96853e5d7e21024bd0e0e8597      3636   

                          review_id  rating  \
0  dd6c3698b614b747d6fbd01e65186e34       0   
1  0e2ba452097b55739c564e1fabcd361e       5   
2  c1c06eeb894e4307875165f4fe8d2f98       4   
3  31a18e28070f4b2cb2139ea92c16c9c5       5   
4  0143ae5b6964035e5b7929b6b27ef92b       5   

                                         review_text  text_reviews_count  \
0                                                  O                 210   
1                my pick for the caldecott so far...                 210   
2  I liked the illustrations, which are are - wel...                  22   
3                    The best book I've read so far.               49850   
4  This book is absolutely enjoyable! I loved the...               49850   

     series language_code is_ebook  average_rating  \
0        []           eng    false            4.05   
1        []           eng    false            4.05   
2        []           eng    false            3.78   
3  [162712]           eng    false            4.12   
4  [162712]           eng    false            4.12   

                                         description  \
0  Xander planned a panda party. Yes, a dandy who...   
1  Xander planned a panda party. Yes, a dandy who...   
2  Now a Storytime from Space Official Selection!...   
3  Twelve-year-old Jonas lives in a seemingly ide...   
4  Twelve-year-old Jonas lives in a seemingly ide...   

                                             authors               publisher  \
0  [{'author_id': '61707', 'role': ''}, {'author_...           Clarion Books   
1  [{'author_id': '61707', 'role': ''}, {'author_...           Clarion Books   
2  [{'author_id': '9860291', 'role': ''}, {'autho...  Dawn Publications (CA)   
3                [{'author_id': '2493', 'role': ''}]                   Ember   
4                [{'author_id': '2493', 'role': ''}]                   Ember   

  num_pages edition_information publication_year  ratings_count  \
0        40                                 2013           1163   
1        40                                 2013           1163   
2        32                                 2016             84   
3       208                                 2006        1311422   
4       208                                 2006        1311422   

                       title       title_without_series  
0       Xander's Panda Party       Xander's Panda Party  
1       Xander's Panda Party       Xander's Panda Party  
2           A Moon of My Own           A Moon of My Own  
3  The Giver (The Giver, #1)  The Giver (The Giver, #1)  
4  The Giver (The Giver, #1)  The Giver (The Giver, #1)

In [23]:
children_df.shape

(5000, 19)

I still have 5000 rows, meaning all the review samples are accounted for, and the columns have increased given that I've combined the information from both DataFrames. Now I'm going to take a closer look at some of the text just to make sure it's there in full (I'll dive into analysis of how the text is represented later on). 

In [24]:
children_df.review_text[1]

'my pick for the caldecott so far...'

In [25]:
children_df.review_text[1000]

'Crenshaw was good, but not the best. Crenshaw was a weird, but funny cat. I thought that the ending was a little cheesy.'

## Children's Literature Genre Info
This is the children's genre subset of data, but the data provided offers multiple possible genre tags for each book, which could be useful looking at potential overlap between categories. Once again, genre info is in a different JSON file, so I'll load that here and follow the same process as with the other DataFrames so far, merging based on shared book IDs.

In [26]:
genres_df = pd.read_json('/Users/ashleyfeiler/Documents/data_science/Goodreads-Genre-Reviews-Analysis/data/goodreads_book_genres_initial.json', lines=True)

In [27]:
genres_df.head()

book_id                                             genres
0  5333265      {'history, historical fiction, biography': 1}
1  1333909  {'fiction': 219, 'history, historical fiction,...
2  7327624  {'fantasy, paranormal': 31, 'fiction': 8, 'mys...
3  6066819  {'fiction': 555, 'romance': 23, 'mystery, thri...
4   287140                                 {'non-fiction': 3}

In [28]:
genres_df.loc[genres_df['book_id'] == 3636]

book_id                                             genres
2094711     3636  {'children': 2091, 'fiction': 1571, 'young-adu...

In [29]:
children_df = pd.merge(children_df, genres_df)
children_df['Genre_Tag'] = 'children'

Here I've merged the list of genres associated with each book, but have also broadcast the genre tag 'children' to every entry in this category to represent that it is part of the children's literature genre subset. This will be helpful for making a DataFrame with all the different genres combined to keep different genres differentiated.

In [30]:
children_df.head()

user_id   book_id  \
0  a2d6dd1685e5aa0a72c9410f8f55e056  17165891   
1  6adb9b5f52e7a89b820d8afcb6329b0e  17165891   
2  686b174e58665fe35cd89c64c0607992  29064279   
3  e5b6484501d2667dddf7df4d588311de      3636   
4  f59779e96853e5d7e21024bd0e0e8597      3636   

                          review_id  rating  \
0  dd6c3698b614b747d6fbd01e65186e34       0   
1  0e2ba452097b55739c564e1fabcd361e       5   
2  c1c06eeb894e4307875165f4fe8d2f98       4   
3  31a18e28070f4b2cb2139ea92c16c9c5       5   
4  0143ae5b6964035e5b7929b6b27ef92b       5   

                                         review_text  text_reviews_count  \
0                                                  O                 210   
1                my pick for the caldecott so far...                 210   
2  I liked the illustrations, which are are - wel...                  22   
3                    The best book I've read so far.               49850   
4  This book is absolutely enjoyable! I loved the...               49850   

     series language_code is_ebook  average_rating  ...  \
0        []           eng    false            4.05  ...   
1        []           eng    false            4.05  ...   
2        []           eng    false            3.78  ...   
3  [162712]           eng    false            4.12  ...   
4  [162712]           eng    false            4.12  ...   

                                             authors               publisher  \
0  [{'author_id': '61707', 'role': ''}, {'author_...           Clarion Books   
1  [{'author_id': '61707', 'role': ''}, {'author_...           Clarion Books   
2  [{'author_id': '9860291', 'role': ''}, {'autho...  Dawn Publications (CA)   
3                [{'author_id': '2493', 'role': ''}]                   Ember   
4                [{'author_id': '2493', 'role': ''}]                   Ember   

  num_pages edition_information publication_year ratings_count  \
0        40                                 2013          1163   
1        40                                 2013          1163   
2        32                                 2016            84   
3       208                                 2006       1311422   
4       208                                 2006       1311422   

                       title       title_without_series  \
0       Xander's Panda Party       Xander's Panda Party   
1       Xander's Panda Party       Xander's Panda Party   
2           A Moon of My Own           A Moon of My Own   
3  The Giver (The Giver, #1)  The Giver (The Giver, #1)   
4  The Giver (The Giver, #1)  The Giver (The Giver, #1)   

                                              genres Genre_Tag  
0  {'children': 143, 'fiction': 15, 'poetry': 9, ...  children  
1  {'children': 143, 'fiction': 15, 'poetry': 9, ...  children  
2  {'children': 13, 'young-adult': 2, 'non-fictio...  children  
3  {'children': 2091, 'fiction': 1571, 'young-adu...  children  
4  {'children': 2091, 'fiction': 1571, 'young-adu...  children  

[5 rows x 21 columns]

## Children's Literature Author Info
So far the children_df DataFrame has everything except author names, which is represented by an author_id deeply embedded in the 'authors' column. The format of the 'authors' column is formatted strangely, so let's take a closer look at the first row's entry.

In [31]:
children_df.authors[0]

[{'author_id': '61707', 'role': ''}, {'author_id': '387985', 'role': 'Illustrations'}]

The information in the 'authors' column is a list of two dictionaries, each with two key:value pairs for the author and illustrator. I only want the first key:value pair of the first dictionary for the author's ID, but it will probably take some digging to get to it. For now, I'm going to load the JSON file with author-specific info.

In [32]:
authors_df = pd.read_json('/Users/ashleyfeiler/Documents/data_science/Goodreads-Genre-Reviews-Analysis/data/goodreads_book_authors.json', lines=True)

In [33]:
authors_df.head()

average_rating  author_id  text_reviews_count              name  \
0            3.98     604031                   7  Ronald J. Fields   
1            4.08     626222               28716     Anita Diamant   
2            3.92      10333                5075    Barbara Hambly   
3            3.68       9212               36262   Jennifer Weiner   
4            3.82     149918                  96     Nigel Pennick   

   ratings_count  
0             49  
1         546796  
2         122118  
3         888522  
4           1740

Now that I've got the author DataFrame, I know I will need to merge based on the author_id, so let's try to extract the author_id from the main DataFrame. I'm not sure how this is going to work, so I'm going to try it on a small sample first.

In [34]:
test_authors = children_df.head()
test_authors.authors.iloc[2][0] 
#First indexes 'authors' entry from  3rd row, then indexes first dicitionary in list

{'author_id': '9860291', 'role': ''}

In [35]:
#For each index in the DataFrame, map the list shown above to one large list
#Extract author id by using .values() to get only the values, cast as a list, and take first value for the author 
test_list = list(test_authors.index.map(lambda x: test_authors.authors.iloc[x][0])) 
list(test_list[0].values())[0]

'61707'

In [36]:
#Apply the forumula above to create a new column with the author_id
test_authors['author_id'] = test_authors.index.map(lambda x: list(test_list[x].values())[0])

/var/folders/h1/bwktpx2129db0qmxbv4w_shw0000gn/T/ipykernel_88529/3428634348.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_authors['author_id'] = test_authors.index.map(lambda x: list(test_list[x].values())[0])


In [37]:
test_authors

user_id   book_id  \
0  a2d6dd1685e5aa0a72c9410f8f55e056  17165891   
1  6adb9b5f52e7a89b820d8afcb6329b0e  17165891   
2  686b174e58665fe35cd89c64c0607992  29064279   
3  e5b6484501d2667dddf7df4d588311de      3636   
4  f59779e96853e5d7e21024bd0e0e8597      3636   

                          review_id  rating  \
0  dd6c3698b614b747d6fbd01e65186e34       0   
1  0e2ba452097b55739c564e1fabcd361e       5   
2  c1c06eeb894e4307875165f4fe8d2f98       4   
3  31a18e28070f4b2cb2139ea92c16c9c5       5   
4  0143ae5b6964035e5b7929b6b27ef92b       5   

                                         review_text  text_reviews_count  \
0                                                  O                 210   
1                my pick for the caldecott so far...                 210   
2  I liked the illustrations, which are are - wel...                  22   
3                    The best book I've read so far.               49850   
4  This book is absolutely enjoyable! I loved the...               49850   

     series language_code is_ebook  average_rating  ...  \
0        []           eng    false            4.05  ...   
1        []           eng    false            4.05  ...   
2        []           eng    false            3.78  ...   
3  [162712]           eng    false            4.12  ...   
4  [162712]           eng    false            4.12  ...   

                publisher num_pages edition_information publication_year  \
0           Clarion Books        40                                 2013   
1           Clarion Books        40                                 2013   
2  Dawn Publications (CA)        32                                 2016   
3                   Ember       208                                 2006   
4                   Ember       208                                 2006   

  ratings_count                      title       title_without_series  \
0          1163       Xander's Panda Party       Xander's Panda Party   
1          1163       Xander's Panda Party       Xander's Panda Party   
2            84           A Moon of My Own           A Moon of My Own   
3       1311422  The Giver (The Giver, #1)  The Giver (The Giver, #1)   
4       1311422  The Giver (The Giver, #1)  The Giver (The Giver, #1)   

                                              genres Genre_Tag author_id  
0  {'children': 143, 'fiction': 15, 'poetry': 9, ...  children     61707  
1  {'children': 143, 'fiction': 15, 'poetry': 9, ...  children     61707  
2  {'children': 13, 'young-adult': 2, 'non-fictio...  children   9860291  
3  {'children': 2091, 'fiction': 1571, 'young-adu...  children      2493  
4  {'children': 2091, 'fiction': 1571, 'young-adu...  children      2493  

[5 rows x 22 columns]

This seems to work, time to do it to all the data!

In [38]:
author_list = list(children_df.index.map(lambda x: children_df.authors.iloc[x][0]))
list(author_list[0].values())[0]

'61707'

In [39]:
children_df['author_id'] = children_df.index.map(lambda x: int(list(author_list[x].values())[0]))
children_df.shape

(5000, 22)

In [40]:
authors_df = authors_df[['author_id', 'name']]
authors_df.head()

author_id              name
0     604031  Ronald J. Fields
1     626222     Anita Diamant
2      10333    Barbara Hambly
3       9212   Jennifer Weiner
4     149918     Nigel Pennick

In [41]:
children_df = pd.merge(children_df, authors_df)
children_df.shape

(5000, 23)

The author data has been merged, the number of rows is still the same, the number of columns has increased to account for the new author data, and the author's name is now included!

## Language Filtering
Given the information about all the different language codes before, I want to filter out only those books written in Enlglish. First lets look at a list of all the different language codes.

In [42]:
list(children_df.language_code.unique())

['eng', '', 'cze', 'en-US', 'spa', 'en-GB', 'ita', 'ind', 'est', 'ara', 'bul', 'fre', 'rus', 'per', 'vie', 'ger', 'gre', 'swe', 'lav', 'slo', 'por', 'dan', 'rum', 'ukr', 'nl', 'fin', 'mul', 'fil', 'en-CA', 'nor', 'lit', 'hun', 'ben', 'afr']

Most of these are 3-letter standardized language codes, but some are unrecognized. I'm going to take a look at the titles of books with codes I'm unsure of to make sure I'm not missing anything.

In [43]:
children_df.loc[children_df['language_code'] == 'mul'].title.iloc[:10] #Unknown?

1665    Viva Frida
Name: title, dtype: object

In [44]:
children_df.loc[children_df['language_code'] == ''].title.iloc[-10:] #Unknown?

4983                         Catching a Storyfish
4985                        Bear and Bee Too Busy
4987    Petal and Poppy and the Mystery Valentine
4990                      My Grandparents Love Me
4991                             Boss of the Pool
4994      Star Wars: The Original Trilogy Stories
4995             You Read to Me, I'll Read to You
4996                                         Boo!
4997                         Mouse Noses on Toast
4998                                 Hello, Snow!
Name: title, dtype: object

And for the languages that are recognized, let's just look at what some non-English languages look like.

In [45]:
children_df.loc[children_df['language_code'] == 'cze'].title.iloc[:10] #Czech

65                                      Dárce (Dárce, #1)
307                                            Malý princ
308                                            Malý princ
365                             Jen jestli si nevymejšlíš
2632                                          Železný muž
4114    Všechna moje strašidla - Tajemství strašidelné...
4668                                                  Noc
Name: title, dtype: object

In [46]:
children_df.loc[children_df['language_code'] == 'fin'].title.iloc[:10] #Finnish

1498                         Kalmankita : Legenda alkaa
2061                      Viirun ja Pesosen joulupuuhat
2327    Lohikäärmerouvan kosto (Lohikäärmeakatemia, #2)
3724                        Mörkövahti (Mörkövahti, #1)
3904                       Mitä sain tietää meduusoista
4023                          Lätkä-Lauri ja ihmeräpylä
4795    Koiramäen Martta ja tiernapojat (Koiramäki, #5)
4989           Tatu ja Patu etsivinä: Tapaus Puolittaja
Name: title, dtype: object

While some of these language codes seem to include English titles, especially the empty string instead of a language code, to keep things consistent, I am going to filter only those books with some form of English as an assigned language code.

In [47]:
english = (children_df.language_code == 'eng') | (children_df.language_code == 'en-GB') | (children_df.language_code == 'en-US') | (children_df.language_code == 'en-CA')
children_df[english].head(3)

user_id   book_id  \
0  a2d6dd1685e5aa0a72c9410f8f55e056  17165891   
1  6adb9b5f52e7a89b820d8afcb6329b0e  17165891   
3  e68c221b98d5561680dd6838f059b55f   6697305   

                          review_id  rating  \
0  dd6c3698b614b747d6fbd01e65186e34       0   
1  0e2ba452097b55739c564e1fabcd361e       5   
3  85b39c5ef18ea1612abe327778594bc1       4   

                                         review_text  text_reviews_count  \
0                                                  O                 210   
1                my pick for the caldecott so far...                 210   
3  This time Dan and Amy go to the Bahamas and Ja...                 554   

     series language_code is_ebook  average_rating  ... num_pages  \
0        []           eng    false            4.05  ...        40   
1        []           eng    false            4.05  ...        40   
3  [172406]           eng    false            3.98  ...       190   

  edition_information publication_year ratings_count  \
0                                 2013          1163   
1                                 2013          1163   
3                                 2010         39904   

                              title              title_without_series  \
0              Xander's Panda Party              Xander's Panda Party   
1              Xander's Panda Party              Xander's Panda Party   
3  Storm Warning (The 39 Clues, #9)  Storm Warning (The 39 Clues, #9)   

                                              genres Genre_Tag author_id  \
0  {'children': 143, 'fiction': 15, 'poetry': 9, ...  children     61707   
1  {'children': 143, 'fiction': 15, 'poetry': 9, ...  children     61707   
3  {'mystery, thriller, crime': 188, 'young-adult...  children     61707   

             name  
0  Linda Sue Park  
1  Linda Sue Park  
3  Linda Sue Park  

[3 rows x 23 columns]

In [48]:
children_df = children_df[english]
children_df.head(3)

user_id   book_id  \
0  a2d6dd1685e5aa0a72c9410f8f55e056  17165891   
1  6adb9b5f52e7a89b820d8afcb6329b0e  17165891   
3  e68c221b98d5561680dd6838f059b55f   6697305   

                          review_id  rating  \
0  dd6c3698b614b747d6fbd01e65186e34       0   
1  0e2ba452097b55739c564e1fabcd361e       5   
3  85b39c5ef18ea1612abe327778594bc1       4   

                                         review_text  text_reviews_count  \
0                                                  O                 210   
1                my pick for the caldecott so far...                 210   
3  This time Dan and Amy go to the Bahamas and Ja...                 554   

     series language_code is_ebook  average_rating  ... num_pages  \
0        []           eng    false            4.05  ...        40   
1        []           eng    false            4.05  ...        40   
3  [172406]           eng    false            3.98  ...       190   

  edition_information publication_year ratings_count  \
0                                 2013          1163   
1                                 2013          1163   
3                                 2010         39904   

                              title              title_without_series  \
0              Xander's Panda Party              Xander's Panda Party   
1              Xander's Panda Party              Xander's Panda Party   
3  Storm Warning (The 39 Clues, #9)  Storm Warning (The 39 Clues, #9)   

                                              genres Genre_Tag author_id  \
0  {'children': 143, 'fiction': 15, 'poetry': 9, ...  children     61707   
1  {'children': 143, 'fiction': 15, 'poetry': 9, ...  children     61707   
3  {'mystery, thriller, crime': 188, 'young-adult...  children     61707   

             name  
0  Linda Sue Park  
1  Linda Sue Park  
3  Linda Sue Park  

[3 rows x 23 columns]

In [49]:
children_df.language_code.value_counts()

eng      2181
en-US     568
en-GB      98
en-CA      11
Name: language_code, dtype: int64

In [50]:
children_df.shape

(2858, 23)

The shape and value_counts confirm that there are now only English-coded books and that the total number of reviews has now been reduced to 2858 - that's still a decent number of reviews, but the number remaining might vary wildly across genres, so we'll see what kind of sampling may be necessary.

## Formatting
There are still a lot of columns in my working DataFrame, so I want to break it down into only the most essential pieces of data.

In [51]:
children_df.columns

Index(['user_id', 'book_id', 'review_id', 'rating', 'review_text',
       'text_reviews_count', 'series', 'language_code', 'is_ebook',
       'average_rating', 'description', 'authors', 'publisher', 'num_pages',
       'edition_information', 'publication_year', 'ratings_count', 'title',
       'title_without_series', 'genres', 'Genre_Tag', 'author_id', 'name'],
      dtype='object')

In [52]:
children_short = children_df[['user_id', 'rating', 'review_text', 'language_code', 'average_rating', 'num_pages', 'publication_year', 'ratings_count', 'title', 'genres', 'Genre_Tag', 'name']]
children_short.columns = ['User_ID', 'Rating', 'Text', 'Language', 'Avg_Rating', 'Pages', 'Pub_Year', 'Ratings_Count', 'Title', 'Genres', 'Category', 'Author']

In [53]:
children_short.head()

User_ID  Rating  \
0  a2d6dd1685e5aa0a72c9410f8f55e056       0   
1  6adb9b5f52e7a89b820d8afcb6329b0e       5   
3  e68c221b98d5561680dd6838f059b55f       4   
4  2f0831cbee2d9e5a02761989b8e9f491       5   
5  686b174e58665fe35cd89c64c0607992       4   

                                                Text Language  Avg_Rating  \
0                                                  O      eng        4.05   
1                my pick for the caldecott so far...      eng        4.05   
3  This time Dan and Amy go to the Bahamas and Ja...      eng        3.98   
4  Loved the excerpts where Julia, the main chara...      eng        3.67   
5  I liked the illustrations, which are are - wel...      eng        3.78   

  Pages Pub_Year  Ratings_Count                             Title  \
0    40     2013           1163              Xander's Panda Party   
1    40     2013           1163              Xander's Panda Party   
3   190     2010          39904  Storm Warning (The 39 Clues, #9)   
4   240     2007           2929                  Project Mulberry   
5    32     2016             84                  A Moon of My Own   

                                              Genres  Category  \
0  {'children': 143, 'fiction': 15, 'poetry': 9, ...  children   
1  {'children': 143, 'fiction': 15, 'poetry': 9, ...  children   
3  {'mystery, thriller, crime': 188, 'young-adult...  children   
4  {'fiction': 122, 'children': 111, 'young-adult...  children   
5  {'children': 13, 'young-adult': 2, 'non-fictio...  children   

            Author  
0   Linda Sue Park  
1   Linda Sue Park  
3   Linda Sue Park  
4   Linda Sue Park  
5  Jennifer Rustgi

Let's rearrange these columns!

In [54]:
children_short = children_short[[ 'Text', 'Rating', 'Title', 'Author', 'Category', 'Genres', 'Language', 'Pages', 'Pub_Year', 'Avg_Rating', 'Ratings_Count', 'User_ID']]

In [55]:
children_short.shape

(2858, 12)

As seen with the language codes, even though there are no NaN values, some language codes were replaced with empty strings, meaning more elements of the DataFrame could be missing without showing up as a null value. While it's not a big deal if a book is missing its number of pages or year of publication, I do want to make sure I'm not missing any review or rating entries. Since the `.info()` shows that there are no null values in the DataFrame, but there are empty spaces, these are likely filled with empty strings. I created a filter to test if any rows had empty strings or strings with one space in the Text or Rating columns.

In [56]:
empty = (children_short.Text != '') & (children_short.Text != ' ') & (children_short.Rating != '') & (children_short.Rating != ' ')
print(children_short[empty].shape)

(2858, 12)


In [57]:
children_short.Rating.value_counts()

5    1064
4     992
3     514
2     143
0     101
1      44
Name: Rating, dtype: int64

Seems like this DataFrame was already ok! But it's good practice just to double check.

In [58]:
children_short = children_short[empty]
print(children_short.shape)

(2858, 12)


Now that entries without a review text or rating have been excluded, it's finally time to save the remaining data. I'm going to pickle each genre's processed DataFrame that I will then be able to combine together in a separate notebook since they will then be a manageable size.

In [59]:
#f2 = open('data/genre_pkls/children_short.pkl', 'wb')
#pickle.dump(children_short, f2)
#f2.close() 

### Plan for Remaining Genres
I will now repeat the process above for the remaining 7 genre files. I will be following all the same steps as with the Children's Literature data, so I will only flash the head of each DataFrame immediately after reading it in to check that it is correct. Following that, I will only double check the shape of the DataFrame rather than flashing the head of the DataFrame every time to condense the notebook since the process is shown to work.

To see the data processing for the rest of the genres, please see the [data_prep](https://github.com/Data-Science-for-Linguists-2023/Goodreads-Genre-Reviews-Analysis/tree/main/data_prep) folder. Splitting the process into multiple notebooks allowed me to save memory, and each notebook follows the exact same process as demonstrated here.

# Poetry Data
## Reviews

In [60]:
#Load poetry review data
poetry_reviews = pd.read_json('/Users/ashleyfeiler/Documents/data_science/Goodreads-Genre-Reviews-Analysis/data/Reviews/excerpts_shuf/goodreads_reviews_poetry_shuf.json', lines=True)
poetry_reviews.head()

user_id   book_id  \
0  d1fb4ab5f06f53929af8ba4e2057bdab  23337735   
1  fe48e05eee1b40da8cb8c981faf023fb   1856017   
2  5aeec2fe761a7fd9be461c38fb78af2f    138371   
3  ca5c77ab48421011ba033eb6b0462a7d   3167946   
4  e2cf04ac7b40875f48871ad97df16c30  18516208   

                          review_id  rating  \
0  30ff5f6e3ee4a24d5a543534a4546fa9       4   
1  ec22e02a40898e14e5d418134030d840       5   
2  fcaa30ac5b28fd833a7522f779d3cb09       4   
3  85fdefee905149a382688c7a8327e89d       5   
4  92ebb47d21495aaf87153aa75f2f3666       5   

                                         review_text  \
0  Recently, poetry has fallen victim to a slew o...   
1                                beautifully organic   
2               A beautiful version of this classic.   
3  One of the things that can be specified when m...   
4  Sidqan mn 'jml rwyt wSf lHb lflsTyny ldhy qr't ..   

                       date_added                    date_updated  \
0  Mon Feb 22 17:22:38 -0800 2016  Thu Feb 25 21:04:39 -0800 2016   
1  Tue Mar 11 17:38:22 -0700 2008  Tue Mar 11 17:39:26 -0700 2008   
2  Tue Jan 03 21:41:28 -0800 2012  Mon Dec 28 02:39:50 -0800 2015   
3  Sat Apr 05 15:24:43 -0700 2008  Sat Apr 05 15:35:05 -0700 2008   
4  Sat Apr 19 14:57:09 -0700 2014  Tue Jun 17 09:27:01 -0700 2014   

                          read_at                      started_at  n_votes  \
0  Thu Feb 25 00:00:00 -0800 2016  Mon Feb 22 00:00:00 -0800 2016        0   
1                                                                        0   
2  Thu Jan 05 00:00:00 -0800 2012  Tue Jan 03 00:00:00 -0800 2012        0   
3                                                                        2   
4  Tue Jun 17 09:27:01 -0700 2014                                        0   

   n_comments  
0           0  
1           0  
2           0  
3           0  
4           1

In [61]:
#Save data sample to share
#poetry_share = poetry_reviews.head()

#f3 = open('data/genre_share/poetry_share.pkl', 'wb')
#pickle.dump(poetry_share, f3)
#f3.close() 

In [62]:
poetry_reviews.shape

(5000, 11)

In [63]:
poetry_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   user_id       5000 non-null   object
 1   book_id       5000 non-null   int64 
 2   review_id     5000 non-null   object
 3   rating        5000 non-null   int64 
 4   review_text   5000 non-null   object
 5   date_added    5000 non-null   object
 6   date_updated  5000 non-null   object
 7   read_at       5000 non-null   object
 8   started_at    5000 non-null   object
 9   n_votes       5000 non-null   int64 
 10  n_comments    5000 non-null   int64 
dtypes: int64(4), object(7)
memory usage: 429.8+ KB


## Book Info

In [64]:
#Load poetry book data
poetry_books = pd.read_json('/Users/ashleyfeiler/Documents/data_science/Goodreads-Genre-Reviews-Analysis/data/Books/goodreads_books_poetry.json', lines=True)
poetry_books.head()

isbn  text_reviews_count series country_code language_code  \
0                               1     []           US           eng   
1  0811223981                   2     []           US                 
2  0374428115                   7     []           US                 
3  0156182890                  12     []           US                 
4  1942004192                   4     []           US           eng   

                                     popular_shelves asin is_ebook  \
0  [{'count': '8', 'name': 'to-read'}, {'count': ...         false   
1  [{'count': '100', 'name': 'to-read'}, {'count'...         false   
2  [{'count': '32', 'name': 'to-read'}, {'count':...         false   
3  [{'count': '554', 'name': 'to-read'}, {'count'...         false   
4  [{'count': '228', 'name': 'to-read'}, {'count'...         false   

   average_rating kindle_asin  ... publication_month edition_information  \
0            3.83              ...                11                       
1            3.83  B00U2WY9U8  ...                 4                       
2            4.38              ...                 7                       
3            3.71  B00IWTRB1W  ...                 3                       
4            5.00              ...                12               First   

  publication_year                                                url  \
0             1887  https://www.goodreads.com/book/show/16037549-v...   
1             2015  https://www.goodreads.com/book/show/22466716-f...   
2             2008  https://www.goodreads.com/book/show/926662.Gro...   
3             1964  https://www.goodreads.com/book/show/926667.The...   
4             2015  https://www.goodreads.com/book/show/29065952-l...   

                                           image_url   book_id ratings_count  \
0  https://images.gr-assets.com/books/1348176637m...  16037549             3   
1  https://images.gr-assets.com/books/1404958407m...  22466716            37   
2  https://s.gr-assets.com/assets/nophoto/book/11...    926662            45   
3  https://images.gr-assets.com/books/1382939971m...    926667           115   
4  https://images.gr-assets.com/books/1455198396m...  29065952             9   

    work_id                                    title  \
0   5212748    Vision of Sir Launfal and Other Poems   
1  41905435                 Fairy Tales: Dramolettes   
2    911665  Growltiger's Last Stand and Other Poems   
3    995066                       The Cocktail Party   
4  49294781          Louder Than Everything You Love   

                      title_without_series  
0    Vision of Sir Launfal and Other Poems  
1                 Fairy Tales: Dramolettes  
2  Growltiger's Last Stand and Other Poems  
3                       The Cocktail Party  
4          Louder Than Everything You Love  

[5 rows x 29 columns]

In [65]:
poetry_books.country_code.value_counts()

US    36514
Name: country_code, dtype: int64

In [66]:
print(poetry_books.language_code.unique())

['eng' '' 'tur' 'per' 'ara' 'en-US' 'fre' 'cze' 'nl' 'gre' 'ger' 'ben'
 'spa' 'ukr' 'mul' 'rum' 'dan' 'nno' 'ind' 'por' 'enm' 'fin' 'ita' 'rus'
 'bul' 'scr' 'lit' 'msa' 'swe' 'slv' 'srp' 'pol' 'en-CA' 'nor' 'hin'
 'en-GB' 'slo' 'kat' 'jpn' 'pes' 'lav' 'cat' 'urd' 'vie' 'tlh' 'est' 'fil'
 'heb' 'tam' 'mar' 'zho' 'grc' 'fao' 'mal' 'hun' 'bos' 'arw' 'lat' 'isl'
 'ang' 'glg' 'pan' 'tgl' 'gmh' 'ira' 'hye' '--' 'tha' 'mon' 'dum' 'sin'
 'kur' 'san' 'peo' 'nob' 'amh' 'aze' 'frm' 'sco' 'kor' 'afr' 'bel' 'nep'
 'dut' 'gla' 'tel' 'dgr' 'ota' 'fro' 'aus' 'sqi' 'mkd' 'snd']


In [67]:
#Merge review and book data on book ID
poetry_df = pd.merge(poetry_reviews, poetry_books, on='book_id')

In [68]:
poetry_df.shape

(5000, 39)

In [69]:
poetry_df.columns

Index(['user_id', 'book_id', 'review_id', 'rating', 'review_text',
       'date_added', 'date_updated', 'read_at', 'started_at', 'n_votes',
       'n_comments', 'isbn', 'text_reviews_count', 'series', 'country_code',
       'language_code', 'popular_shelves', 'asin', 'is_ebook',
       'average_rating', 'kindle_asin', 'similar_books', 'description',
       'format', 'link', 'authors', 'publisher', 'num_pages',
       'publication_day', 'isbn13', 'publication_month', 'edition_information',
       'publication_year', 'url', 'image_url', 'ratings_count', 'work_id',
       'title', 'title_without_series'],
      dtype='object')

## Genre Info

In [70]:
#Merge poetry genre info
poetry_df = pd.merge(poetry_df, genres_df)
poetry_df['Genre_Tag'] = 'poetry'

In [71]:
poetry_df.shape

(5000, 41)

## Author Info

In [72]:
#Isolate author ID
poetry_authors = list(poetry_df.index.map(lambda x: poetry_df.authors.iloc[x][0]))
list(poetry_authors[0].values())[0]

'7238734'

In [73]:
poetry_df['author_id'] = poetry_df.index.map(lambda x: int(list(poetry_authors[x].values())[0]))

In [74]:
poetry_df.authors.iloc[0]

[{'author_id': '7238734', 'role': ''}]

In [75]:
#Merge poetry author info
poetry_df = pd.merge(poetry_df, authors_df)

In [76]:
poetry_df.shape

(5000, 43)

## Language Filtering

In [77]:
#Limit data to only English
english = (poetry_df.language_code == 'eng') | (poetry_df.language_code == 'en-GB') | (poetry_df.language_code == 'en-US') | (poetry_df.language_code == 'en-CA')
poetry_df = poetry_df[english]

In [78]:
poetry_df.shape

(2186, 43)

## Formatting

In [79]:
#Cut/rearrange columns
poetry_short = poetry_df[['user_id', 'rating', 'review_text', 'language_code', 'average_rating', 'num_pages', 'publication_year', 'ratings_count', 'title', 'genres', 'Genre_Tag', 'name']]
poetry_short.columns = ['User_ID', 'Rating', 'Text', 'Language', 'Avg_Rating', 'Pages', 'Pub_Year', 'Ratings_Count', 'Title', 'Genres', 'Category', 'Author']
poetry_short = poetry_short[[ 'Text', 'Rating', 'Title', 'Author', 'Category', 'Genres', 'Language', 'Pages', 'Pub_Year', 'Avg_Rating', 'Ratings_Count', 'User_ID']]

In [80]:
poetry_short.shape

(2186, 12)

In [81]:
#Eliminate empty review data
empty = (poetry_short.Text != '') & (poetry_short.Text != ' ') & (poetry_short.Rating != '') & (poetry_short.Rating != ' ')
print(poetry_short.shape)
print(poetry_short[empty].shape)
poetry_short = poetry_short[empty]

(2186, 12)
(2185, 12)


In [82]:
#Pickle processed data
#f4 = open('data/genre_pkls/poetry_short.pkl', 'wb')
#pickle.dump(poetry_short, f4)
#f4.close() 

# Comics/Graphic Novel Data
## Reviews

In [83]:
#Load comics review data
comics_reviews = pd.read_json('/Users/ashleyfeiler/Documents/data_science/Goodreads-Genre-Reviews-Analysis/data/Reviews/excerpts_shuf/goodreads_reviews_comics_graphic_shuf.json', lines=True)
comics_reviews.head()

user_id   book_id  \
0  81949d8bcf2b4a1fa26e6a9ac23cc9fd     38990   
1  4a4bc340fffb6d2c93bf4c00db771a09    194479   
2  2042876ead7b889de0d9eebd2109e517     79840   
3  cec77035cef250b558adb6cc3a46c6e9  26031154   
4  2423e998cabf65c03837928bf907d996    476909   

                          review_id  rating  \
0  e5fa35c4eb3bba7d442d95ac7e56a248       5   
1  517ad8a9f1747449daa6809e78379bef       4   
2  6136806213e3b38d416d3f70a7ec6b17       1   
3  671e2f680a1fa778017e9b2783fa4a88       4   
4  423869621de2b793c873e0c98f8a6cba       3   

                                         review_text  \
0  Honestly one of the best books I've ever read....   
1  I actually really liked this collection! I rea...   
2  A bit more than 200 hundred pages detailing a ...   
3  So delightful! I loved the incorporation of al...   
4  Fun! Great for my 4 and 6 year old--they love ...   

                       date_added                    date_updated  \
0  Sun Feb 10 10:02:20 -0800 2008  Thu Feb 14 09:33:52 -0800 2008   
1  Wed Oct 26 15:31:01 -0700 2016  Sun Nov 06 10:58:54 -0800 2016   
2  Tue Mar 03 05:42:35 -0800 2015  Tue Mar 03 06:15:59 -0800 2015   
3  Wed Sep 07 09:21:19 -0700 2016  Wed Sep 07 09:22:26 -0700 2016   
4  Mon Nov 02 13:30:33 -0800 2009  Mon Nov 02 13:31:57 -0800 2009   

                          read_at                      started_at  n_votes  \
0  Thu Jun 01 00:00:00 -0700 2006                                        0   
1  Fri Nov 04 00:00:00 -0700 2016  Wed Oct 26 00:00:00 -0700 2016        0   
2  Tue Mar 03 06:15:59 -0800 2015  Tue Mar 03 00:00:00 -0800 2015        2   
3  Wed Sep 07 09:22:26 -0700 2016  Wed Sep 07 00:00:00 -0700 2016        2   
4  Thu Oct 01 00:00:00 -0700 2009                                        0   

   n_comments  
0           0  
1           0  
2           0  
3           0  
4           0

In [84]:
#Save data sample to share
#comics_share = comics_reviews.head()

#f5 = open('data/genre_share/comics_share.pkl', 'wb')
#pickle.dump(comics_share, f5)
#f5.close() 

In [85]:
print(comics_reviews.shape)
print(comics_reviews.info())

(5000, 11)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   user_id       5000 non-null   object
 1   book_id       5000 non-null   int64 
 2   review_id     5000 non-null   object
 3   rating        5000 non-null   int64 
 4   review_text   5000 non-null   object
 5   date_added    5000 non-null   object
 6   date_updated  5000 non-null   object
 7   read_at       5000 non-null   object
 8   started_at    5000 non-null   object
 9   n_votes       5000 non-null   int64 
 10  n_comments    5000 non-null   int64 
dtypes: int64(4), object(7)
memory usage: 429.8+ KB
None


## Book Info

In [86]:
#Load comics book data
comics_books = pd.read_json('/Users/ashleyfeiler/Documents/data_science/Goodreads-Genre-Reviews-Analysis/data/Books/goodreads_books_comics_graphic.json', lines=True)
comics_books.head()

isbn  text_reviews_count                            series  \
0                               1                                []   
1  2205073346                   2                                []   
2                               5  [246830, 362583, 362581, 623032]   
3                               1                                []   
4  0930289765                   6                 [266759, 1096220]   

  country_code language_code  \
0           US                 
1           US           fre   
2           US           eng   
3           US           eng   
4           US         en-US   

                                     popular_shelves        asin is_ebook  \
0  [{'count': '228', 'name': 'to-read'}, {'count'...  B00NLXQ534     true   
1  [{'count': '2', 'name': 'bd'}, {'count': '2', ...                false   
2  [{'count': '493', 'name': 'to-read'}, {'count'...                false   
3  [{'count': '222', 'name': 'to-read'}, {'count'...  B06XKGGSB7     true   
4  [{'count': '20', 'name': 'to-read'}, {'count':...                false   

   average_rating kindle_asin  ... publication_month edition_information  \
0            4.12              ...                                         
1            3.94              ...                 1                       
2            4.28              ...                                         
3            4.05  B06XKGGSB7  ...                                         
4            4.06              ...                11                       

  publication_year                                                url  \
0                   https://www.goodreads.com/book/show/25742454-t...   
1             2016  https://www.goodreads.com/book/show/30128855-c...   
2             2012  https://www.goodreads.com/book/show/13571772-c...   
3                   https://www.goodreads.com/book/show/35452242-b...   
4             1997  https://www.goodreads.com/book/show/707611.Sup...   

                                           image_url   book_id ratings_count  \
0  https://s.gr-assets.com/assets/nophoto/book/11...  25742454             1   
1  https://images.gr-assets.com/books/1462644346m...  30128855            16   
2  https://images.gr-assets.com/books/1333287305m...  13571772            51   
3  https://s.gr-assets.com/assets/nophoto/book/11...  35452242             6   
4  https://images.gr-assets.com/books/1307838888m...    707611            51   

    work_id                                              title  \
0  42749946                              The Switchblade Mamma   
1  50558228                                            Cruelle   
2    102217  Captain America: Winter Soldier (The Ultimate ...   
3  54276229  Bounty Hunter 4/3: My Life in Combat from Mari...   
4    693886                          Superman Archives, Vol. 2   

                                title_without_series  
0                              The Switchblade Mamma  
1                                            Cruelle  
2  Captain America: Winter Soldier (The Ultimate ...  
3  Bounty Hunter 4/3: My Life in Combat from Mari...  
4                          Superman Archives, Vol. 2  

[5 rows x 29 columns]

In [87]:
comics_books.country_code.value_counts()

US    89411
Name: country_code, dtype: int64

In [88]:
#Merge review and book data on book ID
comics_df = pd.merge(comics_reviews, comics_books, on='book_id')

In [89]:
print(comics_df.shape)
print(comics_df.columns)

(5000, 39)
Index(['user_id', 'book_id', 'review_id', 'rating', 'review_text',
       'date_added', 'date_updated', 'read_at', 'started_at', 'n_votes',
       'n_comments', 'isbn', 'text_reviews_count', 'series', 'country_code',
       'language_code', 'popular_shelves', 'asin', 'is_ebook',
       'average_rating', 'kindle_asin', 'similar_books', 'description',
       'format', 'link', 'authors', 'publisher', 'num_pages',
       'publication_day', 'isbn13', 'publication_month', 'edition_information',
       'publication_year', 'url', 'image_url', 'ratings_count', 'work_id',
       'title', 'title_without_series'],
      dtype='object')


## Genre Info

In [90]:
#Merge comics genre info
comics_df = pd.merge(comics_df, genres_df)
comics_df['Genre_Tag'] = 'comics_graphic'

In [91]:
comics_df.shape

(5000, 41)

## Author Info

In [92]:
#Isolate author ID
comics_authors = list(comics_df.index.map(lambda x: comics_df.authors.iloc[x][0]))
list(comics_authors[0].values())[0]

'21982'

In [93]:
comics_df['author_id'] = comics_df.index.map(lambda x: int(list(comics_authors[x].values())[0]))

In [94]:
#Merge comics author info
comics_df = pd.merge(comics_df, authors_df)
print(comics_df.shape)

(5000, 43)


## Langauge Filtering

In [95]:
#Limit data to only English
english = (comics_df.language_code == 'eng') | (comics_df.language_code == 'en-GB') | (comics_df.language_code == 'en-US') | (comics_df.language_code == 'en-CA')
comics_df = comics_df[english]
comics_df.shape

(3509, 43)

## Formatting

In [96]:
#Cut/rearrange columns
comics_short = comics_df[['user_id', 'rating', 'review_text', 'language_code', 'average_rating', 'num_pages', 'publication_year', 'ratings_count', 'title', 'genres', 'Genre_Tag', 'name']]
comics_short.columns = ['User_ID', 'Rating', 'Text', 'Language', 'Avg_Rating', 'Pages', 'Pub_Year', 'Ratings_Count', 'Title', 'Genres', 'Category', 'Author']
comics_short = comics_short[[ 'Text', 'Rating', 'Title', 'Author', 'Category', 'Genres', 'Language', 'Pages', 'Pub_Year', 'Avg_Rating', 'Ratings_Count', 'User_ID']]

In [97]:
comics_short.shape

(3509, 12)

In [98]:
#Eliminate empty review data
empty = (comics_short.Text != '') & (comics_short.Text != ' ') & (comics_short.Rating != '') & (comics_short.Rating != ' ')
print(comics_short.shape)
print(comics_short[empty].shape)
comics_short = comics_short[empty]

(3509, 12)
(3505, 12)


In [99]:
#Pickle processed data
#f6 = open('data/genre_pkls/comics_short.pkl', 'wb')
#pickle.dump(comics_short, f6)
#f6.close() 

# Data Processing Continued

This same format is followed to load the addition 5 genres in separate Jupyter Notebooks in order to manage the computer's available memory. Click here for links to the additional data processing notebooks:
- [Young Adult and Mystery/Thriller/Crime Genres](https://github.com/Data-Science-for-Linguists-2023/Goodreads-Genre-Reviews-Analysis/blob/main/data_prep/YA_Mystery_Data_Prep.ipynb)
- [Fantasy/Paranormal Genre](https://github.com/Data-Science-for-Linguists-2023/Goodreads-Genre-Reviews-Analysis/blob/main/data_prep/Fantasy_Data_Prep.ipynb)
- [History/Biography Genre](https://github.com/Data-Science-for-Linguists-2023/Goodreads-Genre-Reviews-Analysis/blob/main/data_prep/History_Data_Prep.ipynb)
- [Romance Genre](https://github.com/Data-Science-for-Linguists-2023/Goodreads-Genre-Reviews-Analysis/blob/main/data_prep/Romance_Data_Prep.ipynb)